README
Installation instructions are included in this release package under

  - dspace/docs/html/index.html
or
  - dspace/docs/pdf/DSpace-Manual.pdf

DSpace version information can be found in this release package under
 - dspace/CHANGES
or
 - dspace/docs/html/History.html

Documentation for the most recent stable release(s) may be downloaded
or viewed online at
 - http://www.dspace.org/latest-release/
 - http://wiki.duraspace.org/display/DSDOC/

Installation instructions for other versions may be different, so you
are encouraged to obtain the appropriate version of the Documentation
(from the links above or from SVN).

To obtain files from the SVN repository and build, please see:

 - https://scm.dspace.org/svn/repo/dspace/tags/


Please refer any further problems to the dspace-tech@lists.sourceforge.net
mailing list.

 - http://sourceforge.net/mail/?group_id=19984


Detailed Issue Tracking for DSpace is done on our JIRA Issue Tracker

 - http://jira.dspace.org/


To contribute to DSpace, please see:

 - https://wiki.duraspace.org/display/DSPACE/HowToContribute


For more details about DSpace, including a list of service providers,
places to seek help, news articles and lists of other users, please see:

 - http://www.dspace.org/


DSpace source code licensing information available online at:
 - http://www.dspace.org/license/

Copyright (c) 2002-2010, DuraSpace.  All rights reserved.


dspace-semantic-api/changelog.txt
-------------------------------------------------------------------------------
f82449088f74   	DSpace 4.2 with SSv3.0 (SSv3.0)			Aug 03, 2014
-------------------------------------------------------------------------------
f82449088f74 	- moving to dspace-4.2
38e88f9f4e2b 	- (1) Upgrade to owlapi v3.5
		  (2) Upgrade to FaCT++ v1.6.3
		  (3) escape foaf:page URL
5f4ecb1671f9 	- (1) new placeholder texts for several  MS query fields were 
		   added
		  (2) syntactically correct queries with 'no results' are now kept in
		   history
		  (3) height of the ComboBox arrow button was fixed
7f0fe74eacf0 	- (1) warning messages style was changed
		  (2) only successful queries are now kept in history
38cb714d8561 	- new fix for results time
976b75528406 	- restore proper versioning for dspace-semantic-api
e9c9711ea11b 	- hermit moved to 1.3.8.4
881f21b20a42 	- semantic.css contains styles which bring back the css style
		   that ext-all.css overwrites. It generally contais overlays for
		   codemirror.css, ext-all.css, xtheme-gray.css. Styles affect 
		   'semantic.jsp' and 'showIndProperties.jsp'
46756c0c867c 	- moving to dspace 4 (dspace 4.1 with SS and LD facility)
d4d88e08e576 	- (1) ss-specific styles now gathered in semantic.css 
		  (2) extra check for null (properties) Set added 
		  (3) position of DBpedia tooltip is finally correct
be42d841e9e2 	- DBpedia lookup service tooltip
dd72aa36d779 	- (1) enable dbpedia URL injection and RDF 
		  (2) fix results time (from ms to sec)
5b54ef9063bd 	- Jersey implementation for RESTful access to the individual pane
2deeb8ab3078 	- MS syntax highlight (codemirror) + sparql endpoint 
		   (interface only)
-------------------------------------------------------------------------------
84e9eb4ba1c1   	DSpace 3.2 with SSv2.3 (SSv2.3)			Oct 20, 2013
-------------------------------------------------------------------------------
40beedd9a190 	- Properly export XOAI metadata to QDC
32325e882cc9 	- towards dspace-3.x-SSv2.2:
		   (1) fix pom files [versions, declaration 
		   of dependencies] for compatibility with 
		   dspace-3.x
		   (2) add missing dspace.oai.url declaration 
		   inside oai.cfg
		   (3) add the query time printing feature 
		   [new in dspace-3.x] for semantic queries 
		   as well
-------------------------------------------------------------------------------
8f1f42f9e8ba   	DSpace 1.8.2 with SSv2.2 (SSv2.2)		Jul 1, 2013
-------------------------------------------------------------------------------
7960a0e3af41	- Empty webapps folders fix for Git
-------------------------------------------------------------------------------
99893f31d9b6 (svn r268)	- Migrate to Git			Jun 10, 2013
-------------------------------------------------------------------------------
r262	- Update help for "Reload" button 
r251	- SemanticSearchServlet: print the class of generic reasoner exceptions
	- More detailed reasoner version logging
	- Added ontology URL history in options (fixes issue 27)
r234	- Handling of empty autocomplete lists in jsp, in case of exceptions. 	
	- Copying constructor of SU now has private access. 
      	- Enlarge try block in SemanticSearchServlet to handle null SUs. 
          Put showJSP in finally block.                                            
      	- Fixed issue 16.                                                         
-------------------------------------------------------------------------------- 
r219	DSpace 1.8.2 with SSv2.1 (SSv2.1_r2)			Apr 23, 2012
--------------------------------------------------------------------------------
r203	- Update for DSpace 1.8.2 
	- Read oai base URL from “oai.cfg” 
	- HTML encoding for exception messages in the front-end 
	- Selective precomputation of reasoning operations                         
--------------------------------------------------------------------------------
r176	DSpace v1.7.2 with Semantic Search v2.1  (SSv2.1_r1)	Mar 7, 2012
--------------------------------------------------------------------------------
r171	DSpace Semantic API v2.1: 
	- Implemented persistent reasoner registry (Fixes issue 28) 
	- Updated ontology registry 
      	- Semantic Unit keys now depend also on Session IDs and reasoner type 
      	- Added Hermit 1.3.4 support 
      	- Added FaCT++ 1.5.2 support
      	- Updated dependencies in pom. Use berkeleybop repo 
      	- Upgraded to OWL API v3.2.4 
      	- Measure and log reasoner 'instantiation' time 
      	- (Ontology / XSLT changes)
r170	JSPUI updates for Semantic Search v2.1: 
      	- Added 'reload' button in Options 
      	- Various updates to reflect API changes 
      	- xsd:anyURI linking (fixes issue 19)
r169	- dspace-ont namespace change 
      	- re-enable dcterms:isPartOf for isPartOfSeries element
	    
--------------------------------------------------------------------------------
	DSpace v1.7.1 with Semantic Search v2.0  (SSv2.0_r4)	Jul 5, 2011 
--------------------------------------------------------------------------------
r156	- Log ontology URL
r155	- Proper URL encoding of expression parameter
r154	- Handle any kind of exception upon executeQuery
r151	- map isPartOfSeries to dc:isPartOf rather than dcterms:isPartOf which 
	  should be used as an object property only. 
r150	- added semantic query logging 
      	- initialize reasoner 
      	- check session attribute ""resultset"" for nullity 
      	- catch NPEs on query execution
r149	- Add the ability to initialize the reasoner by precomputing inferences 
     	- Do not reset reasoner instance if not necessary
r148	- typo corrected
r147	- fixed alignment for help link
r144	- help for semantic search added
r143	- help link for semantic search added
r126	- We capitalized class names and enabled two additional properties
--------------------------------------------------------------------------------
r122	DSpace v1.7.1 with Semantic Search v2.0  (SSv2.0_r3)	Apr 5, 2011 
--------------------------------------------------------------------------------

dspace-api/src/main/java/org/dspace/app/util/Util.java
/*
 * Util.java
 *
 * Version: $Revision: 4743 $
 *
 * Date: $Date: 2010-02-07 02:35:14 +0200 (dom 07 de feb de 2010) $
 *
 * Copyright (c) 2002-2009, The DSpace Foundation.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * - Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * - Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * - Neither the name of the DSpace Foundation nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */
package org.dspace.app.util;

import java.io.InputStream;
import java.io.IOException;
import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Enumeration;
import java.util.Locale;
import java.util.Properties;

import javax.servlet.http.HttpServletRequest;

import org.apache.log4j.Logger;
import org.dspace.core.Constants;


/**
 * Miscellaneous utility methods
 *
 * @author Robert Tansley
 * @author Mark Diggory
 * @version $Revision: 4743 $
 */
public class Util {
        // cache for source version result
        private static String sourceVersion = null;

        private static Logger log = Logger.getLogger(Util.class);
        
        /**
         * Utility method to convert spaces in a string to HTML non-break space
         * elements.
         *
         * @param s
         *            string to change spaces in
         * @return the string passed in with spaces converted to HTML non-break
         *         spaces
         */
        public static String nonBreakSpace(String s) {
            StringBuffer newString = new StringBuffer();
        
            for (int i = 0; i < s.length(); i++)
            {
                char ch = s.charAt(i);
        
                if (ch == ' ')
                {
                    newString.append("&nbsp;");
                }
                else
                {
                    newString.append(ch);
                }
            }
        
            return newString.toString();
        }

        /**
         * Encode a bitstream name for inclusion in a URL in an HTML document. This
         * differs from the usual URL-encoding, since we want pathname separators to
         * be passed through verbatim; this is required so that relative paths in
         * bitstream names and HTML references work correctly.
         * <P>
         * If the link to a bitstream is generated with the pathname separators
         * escaped (e.g. "%2F" instead of "/") then the Web user agent perceives it
         * to be one pathname element, and relative URI paths within that document
         * containing ".." elements will be handled incorrectly.
         * <P>
         *
         * @param stringIn
         *            input string to encode
         * @param encoding
         *            character encoding, e.g. UTF-8
         * @return the encoded string
         */
        public static String encodeBitstreamName(String stringIn, String encoding) throws java.io.UnsupportedEncodingException {
            // FIXME: This should be moved elsewhere, as it is used outside the UI
            StringBuffer out = new StringBuffer();
        
            final String[] pctEncoding = { "%00", "%01", "%02", "%03", "%04",
                    "%05", "%06", "%07", "%08", "%09", "%0a", "%0b", "%0c", "%0d",
                    "%0e", "%0f", "%10", "%11", "%12", "%13", "%14", "%15", "%16",
                    "%17", "%18", "%19", "%1a", "%1b", "%1c", "%1d", "%1e", "%1f",
                    "%20", "%21", "%22", "%23", "%24", "%25", "%26", "%27", "%28",
                    "%29", "%2a", "%2b", "%2c", "%2d", "%2e", "%2f", "%30", "%31",
                    "%32", "%33", "%34", "%35", "%36", "%37", "%38", "%39", "%3a",
                    "%3b", "%3c", "%3d", "%3e", "%3f", "%40", "%41", "%42", "%43",
                    "%44", "%45", "%46", "%47", "%48", "%49", "%4a", "%4b", "%4c",
                    "%4d", "%4e", "%4f", "%50", "%51", "%52", "%53", "%54", "%55",
                    "%56", "%57", "%58", "%59", "%5a", "%5b", "%5c", "%5d", "%5e",
                    "%5f", "%60", "%61", "%62", "%63", "%64", "%65", "%66", "%67",
                    "%68", "%69", "%6a", "%6b", "%6c", "%6d", "%6e", "%6f", "%70",
                    "%71", "%72", "%73", "%74", "%75", "%76", "%77", "%78", "%79",
                    "%7a", "%7b", "%7c", "%7d", "%7e", "%7f", "%80", "%81", "%82",
                    "%83", "%84", "%85", "%86", "%87", "%88", "%89", "%8a", "%8b",
                    "%8c", "%8d", "%8e", "%8f", "%90", "%91", "%92", "%93", "%94",
                    "%95", "%96", "%97", "%98", "%99", "%9a", "%9b", "%9c", "%9d",
                    "%9e", "%9f", "%a0", "%a1", "%a2", "%a3", "%a4", "%a5", "%a6",
                    "%a7", "%a8", "%a9", "%aa", "%ab", "%ac", "%ad", "%ae", "%af",
                    "%b0", "%b1", "%b2", "%b3", "%b4", "%b5", "%b6", "%b7", "%b8",
                    "%b9", "%ba", "%bb", "%bc", "%bd", "%be", "%bf", "%c0", "%c1",
                    "%c2", "%c3", "%c4", "%c5", "%c6", "%c7", "%c8", "%c9", "%ca",
                    "%cb", "%cc", "%cd", "%ce", "%cf", "%d0", "%d1", "%d2", "%d3",
                    "%d4", "%d5", "%d6", "%d7", "%d8", "%d9", "%da", "%db", "%dc",
                    "%dd", "%de", "%df", "%e0", "%e1", "%e2", "%e3", "%e4", "%e5",
                    "%e6", "%e7", "%e8", "%e9", "%ea", "%eb", "%ec", "%ed", "%ee",
                    "%ef", "%f0", "%f1", "%f2", "%f3", "%f4", "%f5", "%f6", "%f7",
                    "%f8", "%f9", "%fa", "%fb", "%fc", "%fd", "%fe", "%ff" };
        
            byte[] bytes = stringIn.getBytes(encoding);
        
            for (int i = 0; i < bytes.length; i++)
            {
                // Any unreserved char or "/" goes through unencoded
                if ((bytes[i] >= 'A' && bytes[i] <= 'Z')
                        || (bytes[i] >= 'a' && bytes[i] <= 'z')
                        || (bytes[i] >= '0' && bytes[i] <= '9') || bytes[i] == '-'
                        || bytes[i] == '.' || bytes[i] == '_' || bytes[i] == '~'
                        || bytes[i] == '/')
                {
                    out.append((char) bytes[i]);
                }
                else if (bytes[i] >= 0)
                {
                    // encode other chars (byte code < 128)
                    out.append(pctEncoding[bytes[i]]);
                }
                else
                {
                    // encode other chars (byte code > 127, so it appears as
                    // negative in Java signed byte data type)
                    out.append(pctEncoding[256 + bytes[i]]);
                }
            }
            log.debug("encoded \"" + stringIn + "\" to \"" + out.toString() + "\"");
        
            return out.toString();
        }

        /** Version of encodeBitstreamName with one parameter, uses default encoding
         * <P>
         * @param stringIn
         *                input string to encode
         * @return the encoded string
         */
        public static String encodeBitstreamName(String stringIn) throws java.io.UnsupportedEncodingException {
                return encodeBitstreamName(stringIn, Constants.DEFAULT_ENCODING);
         }

        /**
          * Formats the file size. Examples:
          *
          *  - 50 = 50B
          *  - 1024 = 1KB
          *  - 1,024,000 = 1MB etc
          *
          *  The numbers are formatted using java Locales
          *
          * @param in The number to covnert
          * @return the file size as a String
          */
        public static String formatFileSize(double in) {
             // Work out the size of the file, and format appropriatly
             // FIXME: When full i18n support is available, use the user's Locale
             // rather than the default Locale.
             NumberFormat nf = NumberFormat.getNumberInstance(Locale.getDefault());
             DecimalFormat df = (DecimalFormat)nf;
             df.applyPattern("###,###.##");
             if (in < 1024)
             {
                 df.applyPattern("0");
                 return df.format(in) +  " " + "B";
             }
             else if (in < 1024000)
             {
                 in = in / 1024;
                 return df.format(in) + " " + "kB";
             }
             else if (in < 1024000000)
             {
                 in = in / 1024000;
                 return df.format(in) + " " + "MB";
             }
             else
             {
                 in = in / 1024000000;
                 return df.format(in) + " " + "GB";
             }
         }

    /**
     * Obtain a parameter from the given request as an int. <code>-1</code> is
     * returned if the parameter is garbled or does not exist.
     *
     * @param request
     *            the HTTP request
     * @param param
     *            the name of the parameter
     *
     * @return the integer value of the parameter, or -1
     */
    public static int getIntParameter(HttpServletRequest request, String param)
    {
        String val = request.getParameter(param);

        try
        {
            return Integer.parseInt(val.trim());
        }
        catch (Exception e)
        {
            // Problem with parameter
            return -1;
        }
    }

    /**
     * Obtain an array of int parameters from the given request as an int. null
     * is returned if parameter doesn't exist. <code>-1</code> is returned in
     * array locations if that particular value is garbled.
     *
     * @param request
     *            the HTTP request
     * @param param
     *            the name of the parameter
     *
     * @return array of integers or null
     */
    public static int[] getIntParameters(HttpServletRequest request,
            String param)
    {
        String[] request_values = request.getParameterValues(param);

        if (request_values == null)
        {
            return null;
        }

        int[] return_values = new int[request_values.length];

        for (int x = 0; x < return_values.length; x++)
        {
            try
            {
                return_values[x] = Integer.parseInt(request_values[x]);
            }
            catch (Exception e)
            {
                // Problem with parameter, stuff -1 in this slot
                return_values[x] = -1;
            }
        }

        return return_values;
    }

    /**
     * Obtain a parameter from the given request as a boolean.
     * <code>false</code> is returned if the parameter is garbled or does not
     * exist.
     *
     * @param request
     *            the HTTP request
     * @param param
     *            the name of the parameter
     *
     * @return the integer value of the parameter, or -1
     */
    public static boolean getBoolParameter(HttpServletRequest request,
            String param)
    {
        return ((request.getParameter(param) != null) && request.getParameter(
                param).equals("true"));
    }

    /**
     * Get the button the user pressed on a submitted form. All buttons should
     * start with the text <code>submit</code> for this to work. A default
     * should be supplied, since often the browser will submit a form with no
     * submit button pressed if the user presses enter.
     *
     * @param request
     *            the HTTP request
     * @param def
     *            the default button
     *
     * @return the button pressed
     */
    public static String getSubmitButton(HttpServletRequest request, String def)
    {
        Enumeration e = request.getParameterNames();

        while (e.hasMoreElements())
        {
            String parameterName = (String) e.nextElement();

            if (parameterName.startsWith("submit"))
            {
                return parameterName;
            }
        }

        return def;
    }

    /**
     * Gets Maven version string of the source that built this instance.
     * @return string containing version, e.g. "1.5.2"; ends in "-SNAPSHOT" for development versions.
     */
    public static String getSourceVersion()
    {
        if (sourceVersion == null)
        {
            Properties constants = new Properties();

            try
            {
                InputStream cis = Util.class.getResourceAsStream
                    ("/META-INF/maven/org.dspace/dspace-api/pom.properties");

                constants.load(cis);
            }
            catch(Exception e)
            {
                log.error(e.getMessage(),e);
            }

            sourceVersion = constants.getProperty("version", "none");
        }
        return sourceVersion;
    }
}
